File: Signal/src/ViewControllers/DebugUI/DebugUISyncMessages.m
//
//  Copyright (c) 2022 Open Whisper Systems. All rights reserved.
//

#import "DebugUISyncMessages.h"
#import "DebugUIContacts.h"
#import "Signal-Swift.h"
#import <SignalCoreKit/Randomness.h>
#import <SignalCoreKit/SignalCoreKit-Swift.h>
#import <SignalMessaging/Environment.h>
#import <SignalMessaging/ThreadUtil.h>
#import <SignalServiceKit/OWSDisappearingMessagesConfiguration.h>
#import <SignalServiceKit/OWSIdentityManager.h>
#import <SignalServiceKit/OWSReceiptManager.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>
#import <SignalServiceKit/TSCall.h>
#import <SignalServiceKit/TSIncomingMessage.h>
#import <SignalServiceKit/TSThread.h>
#import <SignalUI/OWSTableViewController.h>

#ifdef DEBUG

NS_ASSUME_NONNULL_BEGIN

@implementation DebugUISyncMessages

#pragma mark - Factory Methods

- (NSString *)name
{
    return @"Sync Messages";
}

- (nullable OWSTableSection *)sectionForThread:(nullable TSThread *)thread
{
    NSMutableArray<OWSTableItem *> *items = [@[
        [OWSTableItem itemWithTitle:@"Send Contacts Sync Message"
                        actionBlock:^{ [DebugUISyncMessages sendContactsSyncMessage]; }],
        [OWSTableItem itemWithTitle:@"Send Groups Sync Message"
                        actionBlock:^{ [DebugUISyncMessages sendGroupSyncMessage]; }],
        [OWSTableItem itemWithTitle:@"Send Blocklist Sync Message"
                        actionBlock:^{ [DebugUISyncMessages sendBlockListSyncMessage]; }],
        [OWSTableItem itemWithTitle:@"Send Configuration Sync Message"
                        actionBlock:^{ [DebugUISyncMessages sendConfigurationSyncMessage]; }],
        [OWSTableItem itemWithTitle:@"Send Verification Sync Message"
                        actionBlock:^{ [DebugUISyncMessages sendVerificationSyncMessage]; }],
        [OWSTableItem itemWithTitle:@"Send PNI Identity Request"
                        actionBlock:^{ [self.syncManager sendPniIdentitySyncRequestMessage]; }],
    ] mutableCopy];

    return [OWSTableSection sectionWithTitle:self.name items:items];
}

+ (MessageSenderJobQueue *)messageSenderJobQueue
{
    return SSKEnvironment.shared.messageSenderJobQueue;
}

+ (OWSContactsManager *)contactsManager
{
    return Environment.shared.contactsManager;
}

+ (OWSIdentityManager *)identityManager
{
    return [OWSIdentityManager shared];
}

+ (BlockingManager *)blockingManager
{
    return [BlockingManager shared];
}

+ (OWSProfileManager *)profileManager
{
    return [OWSProfileManager shared];
}

+ (id<SyncManagerProtocol>)syncManager
{
    OWSAssertDebug(SSKEnvironment.shared.syncManager);

    return SSKEnvironment.shared.syncManager;
}

#pragma mark -

+ (void)sendContactsSyncMessage
{
    [self.syncManager syncAllContacts].catchInBackground(^(NSError *error) { OWSLogInfo(@"Error: %@", error); });
}

+ (void)sendGroupSyncMessage
{
    DatabaseStorageAsyncWrite(self.databaseStorage, ^(SDSAnyWriteTransaction *transaction) {
        [self.syncManager syncGroupsWithTransaction:transaction completion:^ {}];
    });
}

+ (void)sendBlockListSyncMessage
{
    [self.blockingManager syncBlockListWithCompletion:^ {}];
}

+ (void)sendConfigurationSyncMessage
{
    [SSKEnvironment.shared.syncManager sendConfigurationSyncMessage];
}

+ (void)sendVerificationSyncMessage
{
    [OWSIdentityManager.shared tryToSyncQueuedVerificationStates];
}

@end

NS_ASSUME_NONNULL_END

#endif

File: Signal/src/ViewControllers/DebugUI/DebugUISyncMessages.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalServiceKit
import SignalUI

#if USE_DEBUG_UI

class DebugUISyncMessages: DebugUIPage {

    let name = "Sync Messages"

    func section(thread: TSThread?) -> OWSTableSection? {
        return OWSTableSection(title: name, items: [
            OWSTableItem(title: "Send Contacts Sync Message",
                         actionBlock: { DebugUISyncMessages.sendContactsSyncMessage() }),
            OWSTableItem(title: "Send Blocklist Sync Message",
                         actionBlock: { DebugUISyncMessages.sendBlockListSyncMessage() }),
            OWSTableItem(title: "Send Configuration Sync Message",
                         actionBlock: { DebugUISyncMessages.sendConfigurationSyncMessage() })
        ])
    }

    // MARK: -

    private static func sendContactsSyncMessage() {
        SSKEnvironment.shared.syncManagerRef.syncAllContacts()
            .catch(on: DispatchQueue.global()) { error in
                Logger.info("Error: \(error)")
            }
    }

    private static func sendBlockListSyncMessage() {
        SSKEnvironment.shared.blockingManagerRef.syncBlockList(completion: { })
    }

    private static func sendConfigurationSyncMessage() {
        SSKEnvironment.shared.syncManagerRef.sendConfigurationSyncMessage()
    }
}

#endif
